controllers/topology/reconcile_state.go

```diff
@@ -34,24 +39,88 @@ func (r *ClusterReconciler) reconcileState(ctx context.Context, current, desired
 	if err := r.reconcileInfrastructureCluster(ctx, current, desired); err != nil {
 		return err
 	}
-
-	// TODO: reconcile control plane
+	// Reconcile desired state of the ControlPlane object.
+	if err := r.reconcileControlPlane(ctx, current, desired); err != nil {
+		return err
+	}
 
 	// Reconcile desired state of the InfrastructureCluster object.
 	if err := r.reconcileCluster(ctx, current, desired); err != nil {
 		return err
 	}
 
-	// TODO: reconcile machine deployments
-
-	return nil
+	// Reconcile desired state of the MachineDeployment objects.
+	return r.reconcileMachineDeployments(ctx, current, desired)
 }
 
 // reconcileInfrastructureCluster reconciles the desired state of the InfrastructureCluster object.
 func (r *ClusterReconciler) reconcileInfrastructureCluster(ctx context.Context, current, desired *clusterTopologyState) error {
 	return r.reconcileReferencedObject(ctx, current.infrastructureCluster, desired.infrastructureCluster)
 }
 
+// reconcileControlPlane works to bring the current state of a managed topology in line with the desired state. This involves
+// updating the cluster where needed.
+func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, current, desired *clusterTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+	// Set a default nil return function for the cleanup operation.
+	cleanup := func() error { return nil }
+
+	// If the ControlPlane requires an InfrastructureMachineTemplate
+	// TODO: Check to ensure this is in the clusterclass - make this a common function.
+	// There's a possible issue with using get here - getControlPlaneState uses a class passed to it.
+	// This function also adds a side effect as it's a network call to an outside component - makes testing much more difficult.
+	if ok, err := r.classRequiresInfrastructureMachineTemplate(ctx, desired); err == nil && ok {
+		cpInfraRef, err := getNestedRef(desired.controlPlane.object, "spec", "machineTemplate", "infrastructureRef")
+		if err != nil {
+			return errors.Wrapf(err, "failed to update the %s object,", desired.controlPlane.infrastructureMachineTemplate.GetKind())
+		}
```

Comment:
```suggestion
		cpInfraRef := objToRef(desired.controlPlane.infrastructureMachineTemplate)
```

Reply:
At this point the the infra ref in the desired control plane object is already set to the infra machine template that is in `desired` state : https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/topology/desired_state.go#L124-L128

Reply:
I've left this the same for the time being - but I think there's an overall question between this and the controlPlane.MachineInfrastructure check around how much  trust we're actually putting into the desired objects as sources of truth.